src/pw/fpga/opencl_utils.c

```diff
@@ -0,0 +1,370 @@
+    exit(0);
+  }
+
+  //  Based on the number of devices get their device ids
+  cl_device_id *dev_ids = (cl_device_id*) malloc(sizeof(cl_device_id) * (*num_devices));
+  status = clGetDeviceIDs(pid, device_type, *num_devices, dev_ids, NULL);
+  if(status != CL_SUCCESS){
+    printf("Query for device ids failed\n");
+    free(dev_ids);
+    exit(0);
+  }
+  return dev_ids;
+}
+
+/******************************************************************************
+ * \brief   returns the list of all devices for the specfic platform
+ * \param   context created using device
+ * \param   array of devices
+ * \param   number of devices found
+ * \param   size of FFT3d
+ * \retval  created program
+ *****************************************************************************/
+cl_program getProgramWithBinary(cl_context context, const cl_device_id *devices, unsigned num_device, int N[3], char *data_path){
+  char bin_path[500];
+  char *binary;
+  char *binaries[num_device];
+
+  size_t bin_size;
+  cl_int bin_status, status;
+
+  strcpy(bin_path, data_path);
+
+  // Get the full path of the binary
+  if( !getBinaryPath(bin_path, N) ){
+    printf("No paths to the binary found\n");
+    return NULL;
+  }
+
+  // Load binary to character array
+  binary = loadBinary(bin_path, &bin_size);
+  if(binary == NULL){
+    printf("Could not load binary\n");
+    free(binary);
+    return NULL;
+  }
+
+  binaries[0] = binary;
+
+  // Create the program.
+  cl_program program = clCreateProgramWithBinary(context, 1, devices, &bin_size, (const unsigned char **) binaries, &bin_status, &status);
+  if (status != CL_SUCCESS){
+    printf("Query to create program with binary failed\n");
+    free(binary);
+    return NULL;
+  }
+  free(binary);
+  return program;
+}
+
+static bool fileTestExists(char* filename){
+  if( access( filename, R_OK ) != -1 ) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+static bool getBinaryPath(char *path, int N[3]){
+
+#ifdef __PW_FPGA_SP
+  char *subpath = "/../fpgabitstream/fft3d/synthesis_sp/";
+#else
+  char *subpath = "/../fpgabitstream/fft3d/synthesis_dp/";
+#endif
+  
+  switch(N[0]){
+    case 16 : 
+      strcat(path, subpath);
+      strcat(path, "syn16/");
+      break;
+
+    case 32 : 
+      strcat(path, subpath);
+      strcat(path, "syn32/");
+      break;
+
+    case 64 : 
+      strcat(path, subpath);
+      strcat(path, "syn64/");
+      break;
+
+    default:
+      printf("Path not found for the size (%d,%d,%d)", N[0], N[1], N[2]);
+      exit(0);
+      break;
+  }
+  // strcat(full_path, specific_path); // transfer specific path to filename 
+  strcat(path, "fft3d");      // Append filename
+  strcat(path, ".aocx");      // Append extension
+  strcat(path, "\0");         // Append extension
+
+  if (!fileTestExists(path)){
+    printf("File not found \n");
+    return false;
+  }
+
+  return true;
+}
+
+static char* loadBinary(char *binary_path, size_t *bin_size){
```

Comment:
I would prefer the interface to be `size_t loadBinary(char *path, char**buf)`, since that's what I see more in APIs, but ok.

Reply:
I see this is a better way to make APIs. Thanks for the suggestion, made this change as well 👍